link/trace_event.go

```diff
@@ -0,0 +1,299 @@
+package link
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strconv"
+	"strings"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/internal"
+	"golang.org/x/sys/unix"
+)
+
+// Getting the terminology right is usually the hardest part. For posterity and
+// for staying sane during implementation:
+//
+// - trace event: Entry under <tracefs>/events. Can be tracepoints or kprobes.
+//   Cannot be closed, as they are static objects. Can be instantiated into
+//   perf events (see below).
+// - tracepoint: A predetermined hook point in the kernel. Exposed as trace
+//   events in (sub)directories under <tracefs>/events. Cannot be instantiated,
+//   closed or removed, it is static.
+// - k(ret)probe: Ephemeral trace events based on entry or exit points of
+//   arbitrary exported kernel symbols. kprobe-based (tracefs) trace events can
+//   be created system-wide by writing to the <tracefs>/kprobe_events file, or
+//   they can be scoped to the current process by creating PMU perf events.
+// - perf event: An object instantiated based on an existing trace event or
+//   kernel symbol. Referred to by fd in userspace.
+//   Exactly one eBPF program can be attached to a perf event. Multiple perf
+//   events can be created from a single trace event. Closing a perf event
+//   stops any further invocations of the attached eBPF program.
+
+var (
+	tracefsPath = "/sys/kernel/debug/tracing"
+
+	// Trace event groups and names must adhere to this set of characters.
```

Comment:
IIRC names cannot start with a number

Reply:
```
root@ubuntu-2gb-nbg1-1:~# echo 'p:test123/myprobe __x64_sys_execve' >> /sys/kernel/debug/tracing/kprobe_events

root@ubuntu-2gb-nbg1-1:~# cat /sys/kernel/debug/tracing/kprobe_events
p:test123/myprobe __x64_sys_execve

root@ubuntu-2gb-nbg1-1:~# echo 'p:test123/123myprobe2 __x64_sys_execve' >> /sys/kernel/debug/tracing/kprobe_events
-bash: echo: write error: Invalid argument
```

but we never do that because we always hardcode the prefix, I don't know if it's worth adapting the regex

Reply:
Huh, interesting find! The prefix is only used for generating the group (the component before the slash), but the same rule seems to apply to both groups and names:

```
# echo 'p:1test/probe __x64_sys_execve' >> /sys/kernel/debug/tracing/kprobe_events
echo: write error: invalid argument
# echo 'p:test/1probe __x64_sys_execve' >> /sys/kernel/debug/tracing/kprobe_events
echo: write error: invalid argument
```

Tracepoint and kprobe groups, names and kernel symbols are matched against this regex. I can't remember why I allowed dashes to appear, as the kernel doesn't seem to allow that. There are no kernel symbols containing dashes either, nor are there any that start with a number. Will remove the dash and disallow leading numbers.